aptos-move/block-executor/src/view.rs

```diff
@@ -1395,37 +1413,70 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> TResourceVi
         state_key: &Self::Key,
         maybe_layout: Option<&Self::Layout>,
     ) -> PartialVMResult<Option<StateValue>> {
-        self.get_resource_state_value_impl(
-            state_key,
-            UnknownOrLayout::Known(maybe_layout),
-            ReadKind::Value,
-        )
-        .map(|res| res.into_value())
+        let start_instant = self.maybe_profiler_state.as_ref().map(|_| Instant::now());
+
+        let ret = self
+            .get_resource_state_value_impl(
+                state_key,
+                UnknownOrLayout::Known(maybe_layout),
+                ReadKind::Value,
+            )
+            .map(|res| res.into_value());
+
+        if let Some(profiler_state) = &self.maybe_profiler_state {
+            profiler_state.borrow_mut().record_resource_view_stat(
+                start_instant.expect("Must be set at the function start"),
+            );
+        }
+
+        ret
     }
 
     fn get_resource_state_value_metadata(
         &self,
         state_key: &Self::Key,
     ) -> PartialVMResult<Option<StateValueMetadata>> {
-        self.get_resource_state_value_impl(state_key, UnknownOrLayout::Unknown, ReadKind::Metadata)
+        let start_instant = self.maybe_profiler_state.as_ref().map(|_| Instant::now());
+
+        let ret = self
+            .get_resource_state_value_impl(state_key, UnknownOrLayout::Unknown, ReadKind::Metadata)
             .map(|res| {
                 if let ReadResult::Metadata(v) = res {
                     v
                 } else {
                     unreachable!("Read result must be Metadata kind")
                 }
-            })
+            });
+
+        if let Some(profiler_state) = &self.maybe_profiler_state {
+            profiler_state.borrow_mut().record_resource_view_stat(
```

Comment:
this is pretty ugly lol, we can follow the timer pattern that prometheus has (record on drop)

Reply:
refactored, went from ugly to maybe beautiful :)